src/tree/tree.rs

```diff
@@ -131,100 +108,21 @@ where
-    where
-        A: PoseidonWalkableAnnotation<NStack<L, A, S>, D, L, S>,
+        start: u64,
+    ) -> Result<impl IntoIterator<Item = Result<&L, CanonError>>, CanonError>
     {
-        PoseidonTreeIterator::new(&self, data)
-    }
-}
-
-/// Main iterator of the poseidon tree.
-///
-/// Depends on an implementation of `PoseidonWalkableAnnotation` for the tree annotation
-///
-/// Every iteration will check for a valid `PoseidonWalkableAnnotation::poseidon_walk` call and
-/// return a next leaf if the provided data holds true for the implemented logic
-///
-/// The data can be any struct that implements `Clone`, and will be used to define the traversal
-/// path over the tree.
-pub struct PoseidonTreeIterator<L, A, S, D, const DEPTH: usize>
-where
-    L: PoseidonLeaf<S>,
-    A: PoseidonTreeAnnotation<L, S>,
-    S: Store,
-    D: Clone,
-{
-    tree: PoseidonTree<L, A, S, DEPTH>,
-    pos: usize,
-    data: D,
-}
-
-impl<L, A, S, D, const DEPTH: usize> PoseidonTreeIterator<L, A, S, D, DEPTH>
-where
-    L: PoseidonLeaf<S>,
-    A: PoseidonTreeAnnotation<L, S>,
-    A: PoseidonWalkableAnnotation<NStack<L, A, S>, D, L, S>,
-    S: Store,
-    D: Clone,
-{
-    /// Iterator constructor
-    pub fn new(
-        tree: &PoseidonTree<L, A, S, DEPTH>,
-        data: D,
-    ) -> Result<Self, &'static str> {
-        let tree = tree.clone();
-
-        // TODO - Naive implementation until iterable branch is implemented
-        // https://github.com/dusk-network/microkelvin/issues/23
-        let pos = <Branch<NStack<L, A, S>, S>>::walk(&tree.inner, |w| {
-            A::poseidon_walk(w, data.clone())
-        })
-        .map_err(|_| "Error fetching the branch!")?
-        .map(|l| l.pos())
-        .unwrap_or(u64::max_value()) as usize;
-
-        Ok(Self { tree, pos, data })
-    }
-}
-
-impl<L, A, S, D, const DEPTH: usize> Iterator
-    for PoseidonTreeIterator<L, A, S, D, DEPTH>
-where
-    L: PoseidonLeaf<S>,
-    A: PoseidonTreeAnnotation<L, S>,
-    A: PoseidonWalkableAnnotation<NStack<L, A, S>, D, L, S>,
-    S: Store,
-    D: Clone,
-{
-    type Item = Result<L, Error<S::Error>>;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        let pos = self.pos;
-        let (pos_p, overflow) = self.pos.overflowing_add(1);
-        if overflow {
-            return None;
-        }
-        self.pos = pos_p;
-
-        match self.tree.get(pos) {
-            // Hack until iterable branch is available
-            // This will prevent the iteration over non-filtered data
-            // https://github.com/dusk-network/microkelvin/issues/23
-            Ok(Some(l)) if A::poseidon_leaf_found(&l, self.data.clone()) => {
-                Some(Ok(l))
-            }
-            Ok(Some(_)) => self.next(),
-            Err(e) => Some(Err(e)),
-            _ => None,
+        let result = self.inner.nth(start);
+        match result {
+            Ok(Some(iter)) => Ok(iter),
+            _ => Err(CanonError::NotFound),
```

Comment:
No differentiation between an actual error in the backend or `CanonError::NotFound`